### PR TITLE
feat(dialog): ts conversion

### PR DESCRIPTION
**Pull request by moog16**
related #4225

### CODE REVIEW

File: packages/mdc-dialog/adapter.ts [hunk 1]
```diff
@@ -21,80 +21,46 @@
  * THE SOFTWARE.
  */
 
-/* eslint no-unused-vars: [2, {"args": "none"}] */
-
 /**
  * Adapter for MDC Dialog. Provides an interface for managing:
```
Comment (acdvorak) [inline packages/mdc-dialog/adapter.ts]: Remove this line

File: packages/mdc-dialog/adapter.ts [hunk 2]
```diff
@@ -21,80 +21,46 @@
...
-  notifyClosing(action) {}
-
-  /**
-   * @param {string} action
-   */
-  notifyClosed(action) {}
+
+interface MDCDialogAdapter {
+  addClass(className: string): void;
+
```
Comment (acdvorak) [inline packages/mdc-dialog/adapter.ts]: Remove blank lines except when grouping similar methods

File: packages/mdc-dialog/foundation.ts [hunk 3]
```diff
@@ -210,45 +184,15 @@ class MDCDialogFoundation extends MDCFoundation {
...
-
-  /**
-   * @param {!Event} evt
-   * @private
-   */
-  handleInteraction(evt) {
+  handleInteraction(evt: MouseEvent | KeyboardEvent) {
     const isClick = evt.type === 'click';
-    const isEnter = evt.key === 'Enter' || evt.keyCode === 13;
+    const isEnter = (evt as KeyboardEvent).key === 'Enter' || (evt as KeyboardEvent).keyCode === 13;
```
Comment (acdvorak) [inline packages/mdc-dialog/foundation.ts]: Extract these checks to separate functions to make the code more readable. Ditto below.

File: packages/mdc-dialog/index.ts [hunk 4]
```diff
@@ -22,66 +22,17 @@
  */
 
 import MDCComponent from '@material/base/component';
+import {closest, matches} from '@material/dom/ponyfill';
 import {MDCRipple} from '@material/ripple/index';
-
+import * as createFocusTrap from 'focus-trap';
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Use two separate imports: one for the function, and one for the type definitions:

```ts
import createFocusTrap from 'focus-trap';
import * as FocusTrapLib from 'focus-trap';
```

File: packages/mdc-dialog/index.ts [hunk 5]
```diff
@@ -111,16 +62,50 @@ class MDCDialog extends MDCComponent {
     this.foundation_.setAutoStackButtons(autoStack);
   }
 
+  static attachTo(root: HTMLElement) {
+    return new MDCDialog(root);
+  }
+
+  private buttonRipples_!: MDCRipple[];
+
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Remove blank lines, except when separating groups of similar variables

File: packages/mdc-dialog/index.ts [hunk 6]
```diff
@@ -111,16 +62,50 @@ class MDCDialog extends MDCComponent {
...
+  private buttons_!: HTMLElement[];
+
+  private defaultButton_!: HTMLElement | null;
+
+  private container_!: HTMLElement;
+
+  private content_: HTMLElement | null = null;
+  private initialFocusEl_: HTMLElement | null = null;
+  private focusTrapFactory_!: (
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Create a type for this function signature

File: packages/mdc-dialog/index.ts [hunk 7]
```diff
@@ -111,16 +62,50 @@ class MDCDialog extends MDCComponent {
...
+  private handleDocumentKeydown_!: SpecificEventListener<'keydown'>;
+  private handleOpening_!: () => void;
+
+  private handleClosing_!: () => void;
+
+  private layout_!: () => void;
+
   initialize(focusTrapFactory = createFocusTrap, initialFocusEl = null) {
-    this.container_ = /** @type {!Element} */ (this.root_.querySelector(strings.CONTAINER_SELECTOR));
+    this.container_ = this.root_.querySelector(strings.CONTAINER_SELECTOR) as HTMLElement;
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Use generic version: `querySelector<HTMLElement>`. Ditto below.

File: packages/mdc-dialog/index.ts [hunk 8]
```diff
@@ -111,16 +62,50 @@ class MDCDialog extends MDCComponent {
...
+    if (!this.container_) {
+      throw new Error(`Dialog component requires a ${strings.CONTAINER_SELECTOR} container element`);
+    }
     this.content_ = this.root_.querySelector(strings.CONTENT_SELECTOR);
     this.buttons_ = [].slice.call(this.root_.querySelectorAll(strings.BUTTON_SELECTOR));
     this.defaultButton_ = this.root_.querySelector(strings.DEFAULT_BUTTON_SELECTOR);
-    this.buttonRipples_ = [];
-    this.focusTrapFactory_ = focusTrapFactory;
+    this.focusTrapFactory_ = focusTrapFactory as unknown as util.focusTrap;
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Remove unnecessary type casts

File: packages/mdc-dialog/index.ts [hunk 9]
```diff
@@ -111,16 +62,50 @@ class MDCDialog extends MDCComponent {
...
+  private handleInteraction_!: SpecificEventListener<'click'|'keydown'>;
+
+  private handleDocumentKeydown_!: SpecificEventListener<'keydown'>;
+  private handleOpening_!: () => void;
+
+  private handleClosing_!: () => void;
+
+  private layout_!: () => void;
+
   initialize(focusTrapFactory = createFocusTrap, initialFocusEl = null) {
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Add types to method signature

File: packages/mdc-dialog/index.ts [hunk 10]
```diff
@@ -142,15 +127,15 @@ class MDCDialog extends MDCComponent {
       document.removeEventListener('keydown', this.handleDocumentKeydown_);
     };
 
-    this.listen('click', this.handleInteraction_);
-    this.listen('keydown', this.handleInteraction_);
+    this.listen('click', this.handleInteraction_ as EventListener);
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Remove unnecessary type cast

File: packages/mdc-dialog/index.ts [hunk 11]
```diff
@@ -111,16 +62,50 @@ class MDCDialog extends MDCComponent {
...
     this.buttons_ = [].slice.call(this.root_.querySelectorAll(strings.BUTTON_SELECTOR));
     this.defaultButton_ = this.root_.querySelector(strings.DEFAULT_BUTTON_SELECTOR);
-    this.buttonRipples_ = [];
-    this.focusTrapFactory_ = focusTrapFactory;
+    this.focusTrapFactory_ = focusTrapFactory as unknown as util.focusTrap;
     this.initialFocusEl_ = initialFocusEl;
+    this.buttonRipples_ = [];
 
-    for (let i = 0, buttonEl; buttonEl = this.buttons_[i]; i++) {
+    for (let i = 0; i < this.buttons_.length; i++) {
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Use `for...of` loop instead

File: packages/mdc-dialog/index.ts [hunk 12]
```diff
@@ -111,16 +62,50 @@ class MDCDialog extends MDCComponent {
...
+    element: HTMLElement | string,
+    userOptions?: createFocusTrap.Options,
+  ) => createFocusTrap.FocusTrap;
+
+  private focusTrap_!: createFocusTrap.FocusTrap;
+
+  private handleInteraction_!: SpecificEventListener<'click'|'keydown'>;
+
+  private handleDocumentKeydown_!: SpecificEventListener<'keydown'>;
+  private handleOpening_!: () => void;
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Use `EventListener` instead (ditto below)

File: packages/mdc-dialog/index.ts [hunk 13]
```diff
@@ -176,33 +161,40 @@ class MDCDialog extends MDCComponent {
...
         if (this.defaultButton_) {
           this.defaultButton_.click();
         }
       },
+      eventTargetMatches: (target, selector) => {
+        if (!target) return false;
+        return matches(target as Element, selector);
+      },
+      getActionFromEvent: (event: MouseEvent | KeyboardEvent) => {
+        if (!event.target) return '';
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]:
```ts
if (!event.target) {
  return '';
}
```

File: packages/mdc-dialog/index.ts [hunk 14]
```diff
@@ -176,33 +161,40 @@ class MDCDialog extends MDCComponent {
...
-        const element = closest(event.target, `[${strings.ACTION_ATTRIBUTE}]`);
-        return element && element.getAttribute(strings.ACTION_ATTRIBUTE);
-      },
       clickDefaultButton: () => {
         if (this.defaultButton_) {
           this.defaultButton_.click();
         }
       },
+      eventTargetMatches: (target, selector) => {
+        if (!target) return false;
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Condense this onto one line:

```ts
eventTargetMatches: (target, selector) => target ? matches(target as Element, selector) : false,
```

File: packages/mdc-dialog/index.ts [hunk 15]
```diff
@@ -176,33 +161,40 @@ class MDCDialog extends MDCComponent {
...
+        return matches(target as Element, selector);
+      },
+      getActionFromEvent: (event: MouseEvent | KeyboardEvent) => {
+        if (!event.target) return '';
+        const element = closest(event.target as Element, `[${strings.ACTION_ATTRIBUTE}]`);
+        return element && element.getAttribute(strings.ACTION_ATTRIBUTE);
+      },
+      hasClass: (className) => this.root_.classList.contains(className),
+      isContentScrollable: () => !!this.content_ && util.isScrollable(/** @type {!Element} */ (this.content_)),
```
Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Remove Closure type cast

Comment (acdvorak) [inline packages/mdc-dialog/index.ts]: Remove `!!`

Comment (kfranqueiro) [reply packages/mdc-dialog/index.ts]: Wouldn't this no longer be boolean without that if `this.content_` doesn't exist? (Though it should really always exist...)

File: packages/mdc-dialog/index.ts (hunk 4 above)
Comment (kfranqueiro) [reply packages/mdc-dialog/index.ts]: Are you suggesting we shouldn't need to re-export FocusTrapLib from our own `./types` (which is where we're currently importing it from)?

File: packages/mdc-dialog/adapter.ts (hunk 1 above)
Comment (acdvorak) [reply packages/mdc-dialog/adapter.ts]: Done

File: packages/mdc-dialog/adapter.ts (hunk 2 above)
Comment (acdvorak) [reply packages/mdc-dialog/adapter.ts]: Done

File: packages/mdc-dialog/index.ts (hunk 6 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Done

File: packages/mdc-dialog/index.ts (hunk 5 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Done

File: packages/mdc-dialog/index.ts (hunk 8 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Turns out we still need the `unknown` type cast. `focus-trap` exports and `.d.ts` definitions are kinda funky.

File: packages/mdc-dialog/index.ts (hunk 7 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Done

File: packages/mdc-dialog/index.ts (hunk 9 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Done

File: packages/mdc-dialog/index.ts (hunk 10 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Done

File: packages/mdc-dialog/index.ts (hunk 11 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Done

File: packages/mdc-dialog/index.ts (hunk 12 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Done

File: packages/mdc-dialog/index.ts (hunk 13 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Done

File: packages/mdc-dialog/index.ts (hunk 14 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Done

File: packages/mdc-dialog/index.ts (hunk 15 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Done

Comment (kfranqueiro) [reply packages/mdc-dialog/index.ts]: Nevermind, I guess this case is handled within `util.isScrollable`.

Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Yeah, I updated `util.isScrollable()` to accept a null `element` argument.

IIRC `content_` can technically be null if you only have a title and action buttons.

File: packages/mdc-dialog/foundation.ts (hunk 3 above)
Comment (acdvorak) [reply packages/mdc-dialog/foundation.ts]: Done

Comment (acdvorak) [reply packages/mdc-dialog/foundation.ts]: Ken wasn't wild about extracting these out into `util` functions, so I reverted that change.

File: packages/mdc-dialog/index.ts (hunk 4 above)
Comment (acdvorak) [reply packages/mdc-dialog/index.ts]: Turns out we need to do it this way:

```ts
import * as createFocusTrap from 'focus-trap';

// Function signature
createFocusTrap(...);

// Interface definitions
createFocusTrap.FocusTrap;
createFocusTrap.Options;
```

Although Focus Trap's `.d.ts` file claims to have a default export, default imports don't work:

```ts
import createFocusTrap from 'focus-trap';

// ERROR: undefined is not a function
createFocusTrap(...);
```

**Pull request merged**